crypto/aes/asm/aes-s390x.pl

```diff
@@ -1405,7 +1405,42 @@ ()
 	clr	%r0,%r1
 	jl	.Lctr32_software
 
-	stm${g}	%r6,$s3,6*$SIZE_T($sp)
+	stm${g}	$s2,$s3,10*$SIZE_T($sp)
+
+	clr	$len,%r1
+	jle	.Lctr32_nokma	# kma is slower for <= 16 blocks
+
+	larl	%r1,OPENSSL_s390xcap_P
+	lr	$s2,%r0
+	llihh	$s3,0x8000
+	srlg	$s3,$s3,0($s2)
+	ng	$s3,S390X_KMA(%r1)	# check kma capability vector
+	jz	.Lctr32_nokma
+
+	aghi	$sp,-112	# prepare parameter block
```

Comment:
I'm uneasy about this. ABI specification kind of says that subroutine has to allocate standard frame *plus* whatever it needs, not vice versa, as implied here. Indeed, it reads "*Except* for the stack frame header and any padding necessary to make the entire frame a multiple of 8 bytes in length, a function need not allocate space for the areas
that it does not use." I do realize that -112 probably works, but I'd be reluctant to make the assumption. Spec also says words about back-chaining stack frames, though with reservation "*if* back chain is implemented." I don't see that "my" gcc does back-chaining, but question is if there is a chance that this module would be used in environment where back chain is implemented. I mean if there is chance, then maybe it would be appropriate to do it anyway, just in case...

Reply:
Reading the relevant parts again, your probably right about this. Anyway. ill check with our compiler folks and rework this.

As a side note, the zseries abi specs from http://refspecs.linuxbase.org/ are abandoned and there will be a revised version on developerworks but its not yet clear when.

Reply:
Well, even if new version will say -112 is ok, I'd say that it would still be appropriate to stick with old way, because it will be compliant with *both* old and new. As implied, by making least assumptions we have better chances to reuse the code.

Reply:
Discussed this with s390 gcc maintainer:
- the abi requires stdframe to be allocated only for a callee (i.e. the code as is conforms to the abi)
- gcc always allocates stdframe if the function needs stack space (i.e. either r15 is not decremented or at least by stdframe). this is not necessary per abi (and it would probably even be good if it didnt do so).

Im kind of neutral to this because it shouldnt affect performance either way. So if youd like the code to mimic what gcc is doing, im fine with that.

Reply:
> the abi requires stdframe to be allocated only for a callee

Arguably that's not how published specification is interpreted. Besides what is already quoted, next sentence is "If a function does not call any other functions and does not require any of the other parts of the stack frame, it need not establish a stack frame." In other words combination of sentences reads "if you need *any* stack, allocate frame with standard header [even in leaf function]." As already said, I'd argue that it's not exclusively about what does most recent spec say, or what works de-facto *today*, but making least assumptions. And since published formulation does not contradict allegedly new one [i.e. "old' way is compatible with "new" one], but not vice versa, it's only appropriate to stick with "old" way. To summarize, please allocate $stdframe+112.

Reply:
okay, so for maximum call compatibility ill also maintain the backchain (and use stdframe only for its default purpose as you suggested) i.e. what gcc would do under -mno-packed-stack and -mbackchain.

Reply:
> gcc -mbackchain.

Oh! It's an option... Well, it only speaks in favour of making least [contradiction-free] assumptions, doesn't it :-) I mean why bother with figuring out what's expected if one can make it work in either case with pair of additional instructions, right?

Reply:
Right. I dont question that making as little assumptions as possible is the way to go.
Rather out of curiosity: what assumptions do you see that were being made in the original submission? I mean the asm subroutines are "black boxes" from compiler's view, they are processed by the assembler (and maybe the preprocessor). So if, after returning from said subroutine, everything is left as required by the abi, what could possibly go wrong? If the abi  specifies (accidentally or not), how a subroutine has to manage its *private* stack (i.e. that is only ever read or written by itself) how can this be a problem for surrounding c code? I can only think in the direction of signal handling, but then again, what could go wrong here?

Reply:
> what could possibly go wrong?

This is a little bit wrong question is my opinion. It's not about *what* can go wrong, but rather which part of the code you'll blame *if* it goes wrong. In other words making least [contradiction-free] assumptions is a way to ensure that you don't have to deal with *any* problem related to calling sequence. It's a way to *not* have to think about what can go wrong, i.e. not having to answer the question. In a way you ask what does it solve, and the answer is it deflects the question itself. In a sense one can argue that it means that it solves nothing, but then you'd have to prove that it won't cause *any* problem, even most unlike and unimaginable. In such case you might as well ask what's harder, to prove, or to arrange things in a way that you don't have to? So what could possibly go wrong? It's argued that we don't even have to think about it...

> what assumptions do you see that were being made in the original submission?

Obviously that non-standard frame won't cause *any* problems, even most unlike and unimaginable. You mentioned that assembly module is to be viewed as a "black box". But by making assumptions like that you effectively "wire" it to specific context, so that in a sense it stops being truly "black" as you can't move it elsewhere as freely. Not without answering question if non-standard frame works even in another context. Well, in a sense one can dismiss all these arguments with "s390x is mono-cultural *enough* to not care about it, there is only one context." Yes, one can. But does one have to? Given that costs of adhering to standard frame are negligible?

Reply:
Okay thanks. I didnt consider "the compiler does not process the asm files and thus cannot depend on what is done inside asm subroutine" an assumption, but i see compiler *could* make assumptions about what is done inside the subroutine from abi requirements and *somehow* depend on that. Otherwise, I already agreed.

Reply:
Just in case for reference, as for "s390x is mono-cultural enough to not care about it". "it" does not refer to s390x platform, but the problem being discussed, i.e. whether or not "non-standard" frame works or not.

Reply:
(At least) I understood what you wanted to say ; -)